### PR TITLE
gh-117142: Port _ctypes to multi-phase init

### PR DESCRIPTION
**Pull request by neonene**
Port `_ctypes` to multi-phase init with the module state enabled.

Module state access:

* Rely on metaclass's mro, which needs to be walked when the C-API metaclass is overridden:

```py
class _swapped_struct_meta(_swapped_meta, type(Structure)):
    pass
class LittleEndianStructure(Structure, metaclass=_swapped_struct_meta):
    ...

# Normal type's mro is less reliable:
type(_SimpleCData)("", (list,), {'_type_': 'i'})
```

| From| access|      | 
| :--- | :--- | :--- | 
| ctype  | `PyType_GetModuleByDef(Py_TYPE(type))` || 
| cdata  | `PyType_GetModuleByDef(Py_TYPE(Py_TYPE(obj)))` || 
| CFieldObject| `_PyType_GetModuleState(Py_TYPE(obj))`| Internal | 
| CThunkObject| `_PyType_GetModuleState(Py_TYPE(obj))`| Internal | 

<br>

* Issue: gh-117142

### CODE REVIEW

**Comment by neonene** [comment]
Changing this to a draft. See a separated PR: #117189.

**Comment by neonene** [comment]
Visiting all `stginfo` members will make the GC more active during a finish, so I propose that the meta `tp_traverse`, `clear` and `dealloc` functions can access a type state, bypassing an unsafe module state. For example:

```c
// from a type assumed to be a ctypes class:
static inline int
_stginfo_from_ctype(PyObject *type, StgInfo **result)
{
    *result = NULL;
    PyTypeObject *PyCType_Type = Py_TYPE(type);
    while (PyCType_Type->tp_base != &PyType_Type) {
        PyCType_Type = PyCType_Type->tp_base;
        if (!PyCType_Type) {
            return 0;
        }
    }
    assert(PyCType_Type->tp_basicsize == Py_TYPE(type)->tp_basicsize);
    assert(PyCType_Type->tp_basicsize > PyType_Type.tp_basicsize);
    StgInfo *info = PyObject_GetTypeData(type, PyCType_Type);
    assert(info != NULL);
    if (!info->initialized) {
        // StgInfo is not initialized. This happens in abstract classes.
        return 0;
    }
    *result = info;
    return 1;
}
```

**Comment by neonene** [comment]
Ready for review.  Visiting more `stginfo` members by #117143 promotes the GC to destroy more ref-cycles during a finish, so I propose that the meta `tp_traverse`, `clear` and `dealloc` functions access a type state without `PyType_GetModuleByDef()`.   An example would be:

```c
// from a type assumed to be a ctypes class:
static inline int
_stginfo_from_ctype(PyObject *type, StgInfo **result)
{
    *result = NULL;
    PyTypeObject *PyCType_Type = Py_TYPE(type)->tp_base;
    while (PyCType_Type->tp_base != &PyType_Type) {
        PyCType_Type = PyCType_Type->tp_base;
        if (!PyCType_Type) {
            return 0;
        }
    }
    assert(PyCType_Type->tp_basicsize == Py_TYPE(type)->tp_basicsize);
    assert(PyCType_Type->tp_basicsize > PyType_Type.tp_basicsize);
    StgInfo *info = PyObject_GetTypeData(type, PyCType_Type);
    assert(info != NULL);
    if (!info->initialized) {
        // StgInfo is not initialized. This happens in abstract classes.
        return 0;
    }
    *result = info;
    return 1;
}
```
```
>python_d -X showrefcount -c "import ctypes"

[2615 refs, 1661 blocks]  # this PR
[175 refs, 132 blocks]    # visit more infos by #117143
[0 refs, 0 blocks]        # apply the example above
```

**Comment by encukou** [comment]
Well, `tp_base` is fine for declaring types in C, but I wouldn't trust it for introspection on arbitrary types, which might be defined in Python, with features like multiple inheritance. (Pity the soul that does that with a `_ctypes` metaclass, but I think it's *possible*.)

We might need a `PyType_GetBaseBySpec`, which would look for the other thing that's `static` -- but we'd need to store a pointer to the spec on types. See also https://github.com/encukou/abi3/issues/19

**Comment by encukou** [comment]
But let's not solve that in *this* PR :)
I'll get to a review later this week.

**Comment by neonene** [comment]
Updated.  Thanks for merging #117143, and thanks to PEP-697, this version looks straightforward, which may be what @erlend-aasland expected in the previous issue.

**Comment by encukou** [comment]
There's one more thing to do: the non-slot methods (`PyMethodDef`) can avoid `PyType_GetModuleByDef` defining class with Argument Clinic (or `METH_METHOD` directly).

Please check https://github.com/neonene/cpython/pull/1. I can do a full PR for this later if you don't beat me to it.

**Comment by neonene** [comment]
OK, I'll try.

**Comment by neonene** [comment]
Applied `METH_METHOD` (PEP-573) as much as possible. Sorry, I just now found your files in my repo.  Indeed, I'm not confident in AC (and git) at all, so feel free to edit this (or I can revert expecting the next PR).

As far as I can see, `c_char_p_from_param()` passes `c_char_p` as the defined class, whose base and meta classes hold a module (C-API).

File: Modules/_ctypes/_ctypes.c [hunk 1]
```diff
@@ -1150,6 +1147,12 @@ size property/method, and the sequence protocol.
 
 */
 
+/*[clinic input]
+class _ctypes.PyCPointerType "PyObject *" "st->PyCPointerType_Type"
```
Comment (erlend-aasland) [inline Modules/_ctypes/_ctypes.c]: FWIW, I've used the practice of defining a "clinic state" macro for getting the module state in generated clinic code. It has proven useful if you want to tweak stuff afterwards (smaller diffs, less churn). You can `git grep clinic_state **/*.c` for inspiration.

Comment (neonene) [reply Modules/_ctypes/_ctypes.c]: IIUC, `*_impl()` should be passed a module state if the AC gets it in advance with some overhead.  Also, `_ctypes` would need a module state getter to be specified in each function clinic input, making a class input have a type name without any operator, **EDIT:** or making the getter return `st`.

Comment (neonene) [reply Modules/_ctypes/_ctypes.c]: Ah, `_PyType_GetModuleState(cls)` or `_PyType_GetModuleState(cls->tp_base)` can be applied to each class. I'll try.

**Comment by neonene** [comment]
Is it acceptable to optimize `PyType_GetModuleByDef` (limited API) by checking the given type before the MRO walk?  This PR passes C-API types to it in almost all cases.  The tuning in `_ctypes` (outside `PyType_GetModuleByDef`) would be fine as well, which should be inlined.

**Comment by encukou** [comment]
> Is it acceptable to optimize PyType_GetModuleByDef (limited API) by checking the given type before the MRO walk?

I seem to recall it had such an optimization at some point. I'm not sure why it's not there now; perhaps we simply went with simpler code in the first PR. If you can measure the speedup, definitely add it :)

File: Modules/_ctypes/_ctypes.c [hunk 2]
```diff
@@ -5610,7 +5799,7 @@ _ctypes_add_objects(PyObject *mod)
         } \
     } while (0)
 
-    ctypes_state *st = GLOBAL_STATE();
+    ctypes_state *st = get_module_state(mod);
```
Comment (encukou) [inline Modules/_ctypes/_ctypes.c:5802]: (note to self: I got to here today; looking good so far!)

File: Modules/_ctypes/callbacks.c [hunk 3]
```diff
@@ -303,7 +303,7 @@ static void closure_fcn(ffi_cif *cif,
                         void *userdata)
 {
     CThunkObject *p = (CThunkObject *)userdata;
-    ctypes_state *st = GLOBAL_STATE();
+    ctypes_state *st = get_module_state_by_class(Py_TYPE(p));
```
Comment (encukou) [inline Modules/_ctypes/callbacks.c:308]: Ah, closures. It gets a bit tricky here.
`get_module_state_by_class` *might* need the GIL held. I didn't check if that's currently necessary, but if not it'd be an implementation detail of e.g. `_PyType_GetModuleState`, which can change at any time.

Please move the `PyGILState_Ensure` and `PyGILState_Release` from `_CallPythonObject` to here. Add something like “needs the GIL” to `_CallPythonObject`.

---

Also, AFAIK `PyGILState_Ensure` is, sadly, incompatible with multiple interpreters, so the `{Py_mod_multiple_interpreters, Py_MOD_PER_INTERPRETER_GIL_SUPPORTED}` slot is premature. We should probably initially disallow creating closures outside the main interpreter.

This PR is very long already. A tricky issue wants a shorter PR, to get more eyes on it. So here, please just make the slot `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` for now.

File: Modules/_ctypes/_ctypes.c [hunk 4]
```diff
@@ -5680,19 +5869,104 @@ _ctypes_mod_exec(PyObject *mod)
...
+
+static void
+module_free(void *module)
+{
+    (void)module_clear((PyObject *)module);
+}
+
+static PyModuleDef_Slot module_slots[] = {
+    {Py_mod_exec, _ctypes_mod_exec},
+    {Py_mod_multiple_interpreters, Py_MOD_PER_INTERPRETER_GIL_SUPPORTED},
```
Comment (encukou) [inline Modules/_ctypes/_ctypes.c]: For now:
```suggestion
    {Py_mod_multiple_interpreters, Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED},
```

**Review by encukou — COMMENTED** [review]
This looks amazing, thank you!
I found one more snag, other than that it looks like we're almost home.

**Comment by neonene** [comment]
Regarding `PyType_GetModuleByDef`, I've applied optimization at #117661, which is not beneficial for the getset of `ctypes`. `_decimal` got around 7% faster in some cases with noises.

File: Modules/_ctypes/callbacks.c (hunk 3 above)
Comment (neonene) [reply Modules/_ctypes/callbacks.c:307]: Moved. Sorry for the delay.

Comment (neonene) [reply Modules/_ctypes/callbacks.c:307]: Ah, wait...

Comment (neonene) [reply Modules/_ctypes/callbacks.c:308]: I added this to the issue's to-do list.

**Review by encukou — APPROVED** [review]
Thank you!

**Comment by erlend-aasland** [comment]
Good job!

**Pull request merged**